pkg/internal/computeresources/tasklevel/tasklevel_test.go

```diff
@@ -0,0 +1,222 @@
+		trs                      v1beta1.TaskRunSpec
+		expectedComputeResources []corev1.ResourceRequirements
+	}{{
+		desc: "only with requests",
+		ts: v1beta1.TaskSpec{
+			Steps: []v1beta1.Step{{
+				Name:    "1st-step",
+				Image:   "image",
+				Command: []string{"cmd"},
+			}, {
+				Name:    "2nd-step",
+				Image:   "image",
+				Command: []string{"cmd"},
+			}},
+		},
+		trs: v1beta1.TaskRunSpec{
+			ComputeResources: &corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{corev1.ResourceCPU: resource.MustParse("2")},
+			},
+		},
+		expectedComputeResources: []corev1.ResourceRequirements{{
+			Requests: corev1.ResourceList{corev1.ResourceCPU: resource.MustParse("1")},
+		}, {
+			Requests: corev1.ResourceList{corev1.ResourceCPU: resource.MustParse("1")},
+		}},
+	}, {
+		desc: "only with limits",
+		ts: v1beta1.TaskSpec{
+			Steps: []v1beta1.Step{{
+				Name:    "1st-step",
+				Image:   "image",
+				Command: []string{"cmd"},
+			}, {
+				Name:    "2nd-step",
+				Image:   "image",
+				Command: []string{"cmd"},
+			}},
+		},
+		trs: v1beta1.TaskRunSpec{
+			ComputeResources: &corev1.ResourceRequirements{
+				Limits: corev1.ResourceList{corev1.ResourceCPU: resource.MustParse("500m")},
+			},
+		},
+		expectedComputeResources: []corev1.ResourceRequirements{{
+			Limits: corev1.ResourceList{corev1.ResourceCPU: resource.MustParse("500m")},
+		}, {
+			Limits: corev1.ResourceList{corev1.ResourceCPU: resource.MustParse("500m")},
+		}},
+	}, {
+		desc: "both with requests and limits",
+		ts: v1beta1.TaskSpec{
+			Steps: []v1beta1.Step{{
+				Name:    "1st-step",
+				Image:   "image",
+				Command: []string{"cmd"},
+			}, {
+				Name:    "2nd-step",
+				Image:   "image",
+				Command: []string{"cmd"},
+			}},
+		},
+		trs: v1beta1.TaskRunSpec{
+			ComputeResources: &corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{corev1.ResourceCPU: resource.MustParse("1")},
+				Limits:   corev1.ResourceList{corev1.ResourceCPU: resource.MustParse("2")},
+			},
+		},
+		expectedComputeResources: []corev1.ResourceRequirements{{
+			Requests: corev1.ResourceList{corev1.ResourceCPU: resource.MustParse("500m")},
+			Limits:   corev1.ResourceList{corev1.ResourceCPU: resource.MustParse("2")},
+		}, {
+			Requests: corev1.ResourceList{corev1.ResourceCPU: resource.MustParse("500m")},
+			Limits:   corev1.ResourceList{corev1.ResourceCPU: resource.MustParse("2")},
+		}},
+	}, {
+		desc: "overwrite StepTemplate resources requirements",
+		ts: v1beta1.TaskSpec{
+			Steps: []v1beta1.Step{{
+				Name:    "1st-step",
+				Image:   "image",
+				Command: []string{"cmd"},
+			}, {
+				Name:    "2nd-step",
+				Image:   "image",
+				Command: []string{"cmd"},
+			}},
+			StepTemplate: &v1beta1.StepTemplate{
+				Resources: corev1.ResourceRequirements{
+					Requests: corev1.ResourceList{
+						corev1.ResourceCPU:    resource.MustParse("500m"),
+						corev1.ResourceMemory: resource.MustParse("500Mi"),
+					},
+				},
+			},
```

Comment:
StepTemplate isn't used in this test at all, so this test isn't testing what we want it to. At the point where `ApplyTaskLevelComputeResources` is called, the step template resource requests have already been applied to the input containers [here](https://github.com/tektoncd/pipeline/blob/e8ee8798a6758bd4bfdd41bb23ec63eff1fb7aec/pkg/pod/pod.go#L138). There are two options here:
1. add a test case where the input steps already have resources on them, and make sure they don't make their way to the output. I don't love this solution because it relies on the behavior I linked to.
2. Create an overall pod test with steptemplate resources and task level resources, and make sure the pod created has the right compute resources. This would be a more thorough test but harder to work with.

I'll let you use your best judgment here but I'll also try to figure out if there's any refactoring I can try here.

Reply:
thanks this context. think good for me mark as a follow-up action.
and will remove this case at this moment.